File: csharp/Behavioral/20-Strategy/Strategy/ContractRateStrategy.cs
namespace Strategy;

// Wraps another strategy and applies an annual-contract discount
public class ContractRateStrategy : IPricingStrategy
{
    private readonly IPricingStrategy _baseStrategy;
    private readonly decimal _discountRate;
    private readonly string _contractRef;

    public ContractRateStrategy(
        IPricingStrategy baseStrategy,
        string contractRef,
        decimal discountRate = 0.15m)
    {
        _baseStrategy = baseStrategy;
        _contractRef  = contractRef;
        _discountRate = discountRate;
    }

    public string Name => $"Contract Rate ({_discountRate:P0} discount)";

    public decimal Calculate(JobCard job) =>
        _baseStrategy.Calculate(job) * (1 - _discountRate);

    public string Explain(JobCard job)
    {
        var baseAmount = _baseStrategy.Calculate(job);
        var discount   = baseAmount * _discountRate;

        return $"{_baseStrategy.Explain(job)} " +
               $"Contract {_contractRef} discount of {_discountRate:P0} " +
               $"(N${discount:N2}) applied.";
    }
}

File: typescript/Behavioral/20-Strategy/Strategy/ContractRateStrategy.ts
import { IPricingStrategy } from "./IPricingStrategy";
import type { JobCard } from "../Model/JobCard";

// Wraps another strategy and applies an annual-contract discount
export class ContractRateStrategy implements IPricingStrategy {
  constructor(
    private readonly baseStrategy: IPricingStrategy,
    private readonly contractRef: string,
    private readonly discountRate: number = 0.15,
  ) {}

  get name(): string {
    return `Contract Rate (${(this.discountRate * 100).toFixed(0)}% discount)`;
  }

  calculate(job: JobCard): number {
    return this.baseStrategy.calculate(job) * (1 - this.discountRate);
  }

  explain(job: JobCard): string {
    const baseAmount = this.baseStrategy.calculate(job);
    const discount = baseAmount * this.discountRate;

    return (
      `${this.baseStrategy.explain(job)} ` +
      `Contract ${this.contractRef} discount of ` +
      `${(this.discountRate * 100).toFixed(0)}% (N$${discount.toFixed(2)}) applied.`
    );
  }
}
